packages/adblocker/src/engine/engine.ts

```diff
@@ -1443,29 +1447,97 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
-          .sort((a, b) => b.getRedirectPriority() - a.getRedirectPriority());
+        const redirectableFilters = this.redirects.matchAll(
+          request,
+          this.isFilterExcluded.bind(this),
+        );
+        const redirectFilters: NetworkFilter[] = [];
+        const removeparamFilters: Map<string, NetworkFilter> = new Map();
+        const removeparamExceptionFilters: Map<string, NetworkFilter> = new Map();
+        for (const filter of redirectableFilters) {
+          if (filter.isRemoveParam()) {
+            if (filter.isException()) {
+              removeparamExceptionFilters.set(filter.removeparam!, filter);
+            } else {
+              removeparamFilters.set(filter.removeparam!, filter);
+            }
+          } else {
+            redirectFilters.push(filter);
+          }
+        }
+
+        // We don't need to match `removeparam` at all if:
+        // * `removeparam` filters not matched
+        // * URL doesn't have search parameter
+        if (removeparamFilters.size !== 0) {
+          redirectUrl = new URL(request.url);
+          if (redirectUrl.searchParams.size !== 0) {
+            for (const [key, filter] of removeparamFilters) {
+              // Remove all params in case of option value is empty
+              // We will not match individual exceptions since it has a higher priority than them
+              if (key === '') {
+                result.filter = filter;
+                removeparamExceptionFilter = removeparamExceptionFilters.get('');
+
+                // In case of non-existence of global exception, we will remove params
+                if (removeparamExceptionFilter === undefined) {
+                  redirectUrl.search = '';
+                }
+
+                break;
+              }
+
+              if (redirectUrl.searchParams.has(key)) {
+                result.filter = filter;
+                removeparamExceptionFilter =
+                  removeparamExceptionFilters.get('') ?? removeparamExceptionFilters.get(key);
+
+                redirectUrl.searchParams.delete(key);
+
+                // Stop if
+                // * the filter was effective
+                // * the filter was excepted with global exception (1st priority)
+                if (
+                  removeparamExceptionFilter === undefined ||
+                  removeparamExceptionFilter.removeparam === ''
+                ) {
+                  break;
+                }
+              }
+            }
+          }
+        }
+
+        // If `result.filter` still remains `undefined`, there
+        // is some extra logic to handle special cases like
+        // redirect-rule and redirect=none.
+        //
+        // * If redirect=none is found, then cancel all redirects.
+        // * Else if redirect-rule is found, only redirect if request would be blocked.
+        // * Else if redirect is found, redirect.
+        if (result.filter === undefined && redirectFilters.length !== 0) {
```

Comment:
What is the rational behind giving priority to removeparam over redirects? Wouldn't it make sense to do it the other way around? A redirect rule is a block rule + a redirect rule (as per uBO semantics). Hence if a request is blocked it should have priority over keeping the request and remove some of the parameters.

Reply:
Fixed by 76e6a5128c1bcb61bced742e2a2868f38bc19c9d